src/herder/test/UpgradesTests.cpp

```diff
@@ -1029,6 +1029,301 @@ TEST_CASE("SCP timing config affects consensus behavior", "[upgrades][herder]")
+        // the previous size (we add no new data here).
+        closeLedger(test.getApp());
+        REQUIRE(test.getLCLSeq() == untilLedger);
+        ensureInMemorySizeIncreased();
+    };
+
+    // We accumulate a small error in the expected state size estimation due
+    // to upgrades. It's tracked in `expectedInMemorySizeDelta` variable.
+    int64_t expectedInMemorySizeDelta =
+        test.getApp().getLedgerManager().getSorobanInMemoryStateSize();
+    auto getExpectedInMemorySize = [&]() {
+        LedgerSnapshot ls(test.getApp());
+        auto res = expectedInMemorySizeDelta;
+        for (auto const& key : addedKeys)
+        {
+            auto le = ls.load(key);
+            res += ledgerEntrySizeForRent(le.current(),
+                                          xdr::xdr_size(le.current()), 23,
+                                          test.getNetworkCfg());
+        }
+        return res;
+    };
+
+    auto getStateSizeWindow = [&]() {
+        LedgerSnapshot ls(test.getApp());
+        LedgerKey key(CONFIG_SETTING);
+        key.configSetting().configSettingID =
+            ConfigSettingID::CONFIG_SETTING_LIVE_SOROBAN_STATE_SIZE_WINDOW;
+        auto le = ls.load(key);
+        REQUIRE(le);
+        std::vector<uint64_t> windowFromLtx =
+            le.current().data.configSetting().liveSorobanStateSizeWindow();
+        auto const& cfg = test.getNetworkCfg();
+        std::vector<uint64_t> windowInMemory(
+            cfg.mSorobanStateSizeSnapshots.begin(),
+            cfg.mSorobanStateSizeSnapshots.end());
+        REQUIRE(windowFromLtx == windowInMemory);
+        return windowFromLtx;
+    };
+    auto getAverageStateSize = [&]() {
+        auto window = getStateSizeWindow();
+        uint64_t sum = 0;
+        for (auto v : window)
+        {
+            sum += v;
+        }
+        uint64_t averageFromWindow = sum / window.size();
+        auto const& cfg = test.getNetworkCfg();
+        uint64_t averageFromConfig = cfg.getAverageSorobanStateSize();
+        REQUIRE(averageFromConfig == averageFromWindow);
+        return averageFromConfig;
+    };
+
+    auto verifyAverageStateSize = [&](uint64_t minSize, uint64_t maxSize) {
+        auto average = getAverageStateSize();
+        if (minSize == maxSize)
+        {
+            REQUIRE(average == maxSize);
+        }
+        else
+        {
+            REQUIRE(average > minSize);
+            REQUIRE(average < maxSize);
+        }
+    };
+
+    auto verifyExpectedInMemorySize = [&](int64_t maxDiff) {
+        int64_t diff =
+            static_cast<int64_t>(test.getApp()
+                                     .getLedgerManager()
+                                     .getSorobanInMemoryStateSize()) -
+            static_cast<int64_t>(getExpectedInMemorySize());
+        if (maxDiff >= 0)
+        {
+            REQUIRE(diff >= 0);
+        }
+        else
+        {
+            REQUIRE(diff <= 0);
+        }
+
+        REQUIRE(diff <= maxDiff);
```

Comment:
I think `REQUIRE(diff <= maxDiff);` is an incorrect check when maxDiff is negative.

Reply:
Fixed.